### ISSUE TITLE
LOAD DATA LOCAL INFILE vs INSERT INTO

### ISSUE THREAD

**Issue by krlmlr** [r-dbi/RMariaDB#11]
<a href="https://github.com/szilard"><img src="https://avatars2.githubusercontent.com/u/844827?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [szilard](https://github.com/szilard)**
_Saturday May 30, 2015 at 03:22 GMT_
_Originally opened as https://github.com/rstats-db/RMySQL/issues/71_

----

For dbWriteTable:
- Latest CRAN version (0.10.3) uses LOAD DATA LOCAL INFILE https://github.com/rstats-db/RMySQL/blob/april/R/table.R#L139
- Latest dev uses INSERT INTO https://github.com/rstats-db/RMySQL/blob/master/R/table.R#L105

Most analytical/columnar/MPP databases (e.g. MySQL-based Infobright) strongly prefer "bulk loads", but my guess is that standard MySQL is faster with LOAD than with INSERT as well http://dev.mysql.com/doc/refman/5.0/en/insert-speed.html

Also, the latest dev uses one INSERT per row rather then "bulk insert" (not to be confused with "bulk load") http://stackoverflow.com/questions/7977172/mysql-multiple-insert-performance Looping 1 row at a time is probably the slowest solution.

Any comments @jeroenooms @hadley on why LOAD was changed to INSERT?

PS: The way I found this is that the latest CRAN 0.10.3 does not work with Infobright (call it a "bug") as Infobright has different FIELDS TERMINATED BY defaults. Adding `FIELDS TERMINATED BY '\t'` here https://github.com/rstats-db/RMySQL/blob/april/R/table.R#L141 would fix the issue on 0.10.3, but that's moot in dev (as LOAD was changed to INSERT).

**Comment by krlmlr** [comment]
<a href="https://github.com/jeroen"><img src="https://avatars0.githubusercontent.com/u/216319?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jeroen](https://github.com/jeroen)**
_Monday Jun 01, 2015 at 10:16 GMT_

----

We can still do more maintenance releases in the 10.x (stable) branch. Can you submit a PR for the [stable](https://github.com/rstats-db/RMySQL/tree/stable) branch with the fix?

**Comment by krlmlr** [comment]
<a href="https://github.com/szilard"><img src="https://avatars2.githubusercontent.com/u/844827?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [szilard](https://github.com/szilard)**
_Monday Jun 01, 2015 at 20:03 GMT_

----

submitted PR (the brach is named "april" not "stable", right?)

**Comment by krlmlr** [comment]
<a href="https://github.com/jeroen"><img src="https://avatars0.githubusercontent.com/u/216319?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jeroen](https://github.com/jeroen)**
_Monday Jun 01, 2015 at 20:11 GMT_

----

Thanks. Either one will do. Can you add a bullet to the `NEWS.md` file ?

**Comment by krlmlr** [comment]
<a href="https://github.com/szilard"><img src="https://avatars2.githubusercontent.com/u/844827?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [szilard](https://github.com/szilard)**
_Monday Jun 01, 2015 at 20:18 GMT_

----

done

**Comment by krlmlr** [comment]
<a href="https://github.com/jeroen"><img src="https://avatars0.githubusercontent.com/u/216319?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jeroen](https://github.com/jeroen)**
_Monday Jun 01, 2015 at 20:28 GMT_

----

Ok merged. FYI I renamed the `april` branch to `stable`.

**Comment by krlmlr** [comment]
<a href="https://github.com/szilard"><img src="https://avatars2.githubusercontent.com/u/844827?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [szilard](https://github.com/szilard)**
_Sunday Jul 05, 2015 at 04:10 GMT_

----

So, this fix above affected only the minor bug in some MySQL based systems (Infobright) in the stable branch. 

The original question of changing `LOAD DATA` to `INSERT` in the dev version still remains. @jeroenooms @hadley I just tested it, the dev version is 100x slower in inserting data from R to MySQL (stable: 1mil rows 11sec, dev: 10K rows 11sec: https://github.com/szilard/benchm-R-mysql )

**Comment by jtelleriar** [comment]
LOAD DATA LOCAL INFILE shall be the fastest way, yes.

Otherwise, a commit shall be executed with each insert when using a transactional storage engines.

**Comment by jtelleriar** [comment]
Maybe an optional clause could be added which allows to choose between several insert statements with their commits, or a single Load Data Infile, which is faster

**Comment by jtelleriar** [comment]
What shall be taken care is that a Commit is not done in InnoDB with each INSERT, because of being too slow.

Sometimes I use dbWriteTable with a MEMORY Engine Table, and then, I convert it to a INNODB table in order to perform the inserts fast, although this cannot be done in computers with little RAM.

**Comment by jtelleriar** [comment]
Information on the topic:
https://mariadb.com/kb/en/library/how-to-quickly-insert-data-into-mariadb/

**Comment by amstilp** [comment]
I'm also seeing ~100x slower performance when running `dbWriteTable` using `RMariaDB` instead of `RMySQL`. This is the one thing stopping me from switching from `RMySQL` to `RMariaDB` since it already takes a long time for my code to load data into tables with `RMySQL`.

Here's some example code to run:
```r
n <- 10000

db_rmysql <- DBI::dbConnect(RMySQL::MySQL(), default.file = "~/.my.cnf")
db_rmariadb <- DBI::dbConnect(RMariaDB::MariaDB(), default.file = "~/.my.cnf")

dat <- data.frame(s = stringi::stri_rand_strings(n, 5))
dim(dat)

query <- paste(
    "CREATE TABLE test (",
    "id INT(11) NOT NULL AUTO_INCREMENT,",
    "s VARCHAR(10) NOT NULL,",
    "PRIMARY KEY (id)",
    ")",
    "ENGINE  =  InnoDB;"
 )
DBI::dbExecute(db_rmariadb, query)

# Check timings for RMySQL
system.time(RMySQL::dbWriteTable(db_rmysql, "test", dat, append = TRUE, row.names = FALSE))
DBI::dbGetQuery(db_rmysql, "TRUNCATE test")

# Check timings for RMariaDB
system.time(RMariaDB::dbWriteTable(db_rmariadb, "test", dat, append = TRUE, row.names = FALSE))
DBI::dbExecute(db_rmariadb, "TRUNCATE test")

DBI::dbExecute(db_rmariadb, "DROP TABLE test")

DBI::dbDisconnect(db_rmysql)
DBI::dbDisconnect(db_rmariadb)
```

For the RMySQL connection, I see the following times on my system:
```
> system.time(RMySQL::dbWriteTable(db_rmysql, "test", dat, append = TRUE, row.names = FALSE))
   user  system elapsed 
  0.025   0.001   0.075 
```
compared with RMariaDB:
```
> system.time(RMariaDB::dbWriteTable(db_rmariadb, "test", dat, append = TRUE, row.names = FALSE))
   user  system elapsed 
  0.404   0.400   8.220 
```

**Comment by krlmlr** [comment]
`dbWriteTable()` tries to be consistent, it's unlikely we can switch the loading method here. Can you use `dbCreateTable()` to create the structure and then manually issue a `LOAD DATA LOCAL INFILE` query?

Perhaps _arkdb_ could be a place where faster import modes could be supported?

**Comment by amstilp** [comment]
Unfortunately, in my real-world use case, I'm loading data into a table that exists and has some data in it already (so `append = TRUE`). I ended up writing my own dbWriteTable function that uses some of the internal `RMariaDB` functions and `LOAD DATA LOCAL INFILE`.

**Comment by krlmlr** [comment]
No luck with the [_arkdb_ package](https://cran.r-project.org/web/packages/arkdb/index.html)?

**Comment by krlmlr** [comment]
Commenting here, because this seems to be the oldest issue mentioning the problem.

RMariaDB uses parameterized statements with `dbBind()` for inserting records. This is very slow, because after each record we need to wait for a response from the server.

MariaDB has a bulk load API: https://mariadb.com/resources/blog/mariadb-bulk-load-api/. First experiments are promising, we could use that to achieve decent gains. However:

- It's not supported on MySQL
- Its use is limited for statements, queries are not supported: https://stackoverflow.com/a/61806932/946850
- This involves a bit of fiddling with translating the various data types

There remain two alternatives: `INSERT INTO ... VALUES (...), (...), (...)` to insert multiple rows at a time, or `LOAD DATA LOCAL INFILE`. I'm leaning towards the second solution:

- This seems to work well enough for RMySQL
- The first approach is limited by the query size: https://stackoverflow.com/q/16335011/946850

We need to take care of a few things:

- Quoting: does it work in all cases, including embedded newlines and other nasty things?
- Character encoding: can be specified with `LOAD DATA`
- BLOBs: use this elegant solution with `UNHEX()`, https://stackoverflow.com/a/5501841/946850
- Time zones

It still looks like a decent amount of work, but gives us the best bang for the buck.

**Comment by krlmlr** [comment]
Benchmark: https://github.com/r-dbi/RMariaDB/issues/104#issuecomment-413345375

**Comment by krlmlr** [comment]
Detailed benchmarks of various operations: #125

**Comment by krlmlr** [comment]
Another benchmark: #162

**Comment by krlmlr** [comment]
`LOAD DATA LOCAL INFILE` is off by default in MySQL 8.0. The functionality will live in a new `mariadbAppendTable()` function, the default in `dbWriteTable()` will keep using the slow but working approach.

**Comment by krlmlr** [comment]
This puts us back to square one: should we perhaps send the table data as one very long SQL query (without parameters)?

**Comment by daniel-hofmann-idnow** [comment]
> 
> 
> `LOAD DATA LOCAL INFILE` is off by default in MySQL 8.0. The functionality will live in a new `mariadbAppendTable()` function, the default in `dbWriteTable()` will keep using the slow but working approach.

Hi,

thanks for your effort. Sorry if I am asking the obvious but I don't understand it completely: Is this 'mariadbAppendTable()` available in Release 1.1.2 or didn't it make it into the release? I would really appreciate it being available. 

BR,
Daniel

**Comment by mcol** [comment]
@daniel-hofmann-idnow  Nothing is available yet, see #223 if you want to track progress.

@krlmlr Is there a way to programmatically check from R if `LOAD DATA LOCAL INFILE` is on or off?

**Comment by krlmlr** [comment]
Yes, it seems we can query this capability: https://dev.mysql.com/doc/refman/8.0/en/load-data-local-security.html . I'd like to leave the responsibility to the user though.

**Comment by krlmlr** [comment]
Closed in #223. I went for a new argument to `dbConnect()` to take advantage of the tests in DBItest.
